### ISSUE TITLE
AutoLoginGuard appears to cause some sort of infinite loop.

### ISSUE THREAD

**Issue by JerrettDavis** [damienbod/angular-auth-oidc-client#1052]
**Describe the bug**
Attempting to access the root route (`/`) with `AutoLoginGuard` present intermittently causes an infinite login loop that crashes the browser.

**To Reproduce**
1. Secure all routes with `AutoLoginGuard`
2. Sign out of the identity server
3. Open console
4. Navigate to web app
5. You may be redirected to the identity server the first time, but navigate back to the webapp.
6. You should start to see a login loop.

**Expected behavior**
Expected library to login user and to return them back to the app homepage.

**Screenshots**
![image](https://user-images.githubusercontent.com/2610199/115294939-d33ccc80-a11e-11eb-92ab-09a9104251fe.png)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome, Chromium Edge
 - Version: 90

**Additional context**
You can force this behavior using the `sample-code-flow-auto-login` sample. Just add `AutoLoginGuard` to the home route, and you should get the loop after you sign in to the app. 

Our end-goal is to have the webapp automatically sign the user in as soon as they access the app. We don't want any sort of landing page. Prior to the existence of the `AutoLoginGuard`, we managed to force this behavior using an approach similar to the `sample-implicit-flow-google` demo (though we needed to change the localstorage key name as `redirect` was being used by the library and causing issues.

**Comment by FabianGosebrink** [comment]
Hey thanks for the issue.

Have you checked that you _do not_ have the `checkAuth` call when using the auto login guard at the home route?

[docs](https://github.com/damienbod/angular-auth-oidc-client/blob/main/docs/features.md#auto-login)

**Comment by JerrettDavis** [comment]
Thanks for the prompt reply @FabianGosebrink! I am actually using `checkAuth` in various services, should those be replaced with `isAuthenticated$` to prevent the login attempts?

**Comment by FabianGosebrink** [comment]
This should work, otherwise, it is a bug. The `checkAuth()` method is gonna be called from the `AuthGuard` and then all values should be set. Let us know if it works.

**Comment by inloox-dev** [comment]
Wir erleichtern Ihnen den Einstieg mit InLoox Planungsvorlagen und stellen die Offline-Funktion detailliert vor.
InLoox Fehler wurde zur Nachverfolgung erstellt
[https://gallery.mailchimp.com/6fb276e204c3e429e34db0e30/images/43782466-0ad5-4511-953f-282e94d647e5.jpg] <https://www.inloox.de/?link=9671>
Guten Tag,


Fehler 25246 wurde zur Nachverfolgung erstellt

Titel: [damienbod/angular-auth-oidc-client] AutoLoginGuard appears to cause some sort of infinite loop. (#1052)

Herzliche Grüße
Dein InLoox Dev-Team

Fehler anzeigen<https://inloox.visualstudio.com//InLoox%20PM/_workitems/edit/25246>
FOLGEN SIE UNS!
[Twitter]<https://twitter.com/inloox_gmbh>
[Facebook]<https://www.facebook.com/InLoox>
[Newsletter-Anmeldung]<https://www.inloox.de/about-us/contact-us/?mode=newsletter>
[Google Plus]<https://plus.google.com/+InlooxDe/posts>
[YouTube]<https://www.youtube.com/user/InLoox>
[InLoox auf Instagram]<https://www.instagram.com/inlooxgmbh/>
Impressum<https://www.inloox.de/impressum/>                             Datenschutzbestimmungen<https://www.inloox.de/datenschutzbestimmungen/>
Copyright © 2019 InLoox GmbH, All rights reserved.

________________________________

**Comment by JerrettDavis** [comment]
I've been unable to test the solution in our own codebase due to other backlog items, but it did correct the issue with the demo project, so I have high hopes. 

I'll go ahead and close the issue for the time being and reopen it if necessary. 

Thanks for the help.

**Comment by FabianGosebrink** [comment]
Perfect, just let us know if the problem still occurs and if yes we will get that going and help you to solve it :), thanks!

**Comment by grffn** [comment]
@FabianGosebrink 
I think this issues occured for me. No `checkAuth()` calls prior to `AuthLoginGuard`. My investigation showed me that loop is occuring because after checking authentication it reads previously stored redirect path (saved before the first redirect to identity provider) and goes there. In my case the path is guarded by `AuthLoginGuard`, so the cycle repeats, and repeats, and repeats

**Comment by FabianGosebrink** [comment]
Do you have a sample repo to reproduce the issue? You can take a config from our samples.

**Comment by grffn** [comment]
I'll try to provide one

**Comment by mrob1600** [comment]
I'm seeing the same looping issue when using AutoLoginGuard on the default route and not calling checkAuth. 

I changed the sample "sample-code-flow-auto-login" to use AutoLoginGuard on the default route and removed the call to checkAuth. For some reason it still works fine (no looping). 

The only differences I see between my app and the sample at this point are I'm using Azure AD as the IdP and using config over http. 

I noticed the app created a local storage variable named redirect as shown below. The value doesn't look like something the redirect should be set to but I could be wrong. 

This is my config and routing:

const setupAction$ = httpClient.get<any>('/api/azuread-client-settings').pipe(
    map((authConfig) => {
      return {
        stsServer: authConfig.stsServer,
        clientId: authConfig.clientID,
        redirectUrl: window.location.origin,
        postLogoutRedirectUri: window.location.origin,
        scope: 'openid profile email api://98d5e485-770f-47c2-a836-35c8cd90e103/access_as_user',
        responseType: 'code',
        silentRenew: true,
        triggerAuthorizationResultEvent: true,
        startCheckSession: false,       
        maxIdTokenIatOffsetAllowedInSeconds: 600,   
        logLevel: LogLevel.Debug,
        historyCleanupOff: true,
        silentRenewUrl: window.location.origin + '/silent-renew.html'
      };
    }),

const appRoutes: Routes = [
  {
    path: '', redirectTo: 'home', pathMatch: 'full'
  },
  {
    path: 'home', component: HomeComponent, canActivate: [AutoLoginGuard]
}


No calls to checkAuth within the application


Recurring within the looping is also this 401:
​ GET https://graph.microsoft.com/oidc/userinfo 401 (Unauthorized)


![LocalStorageRedirect](https://user-images.githubusercontent.com/7670265/116600206-0c332900-a8f7-11eb-8730-434255820d7a.PNG)

**Comment by grffn** [comment]
[sample-code-flow-auto-login.zip](https://github.com/damienbod/angular-auth-oidc-client/files/6404266/sample-code-flow-auto-login.zip)
@FabianGosebrink 
So I toyed a bit with auto login sample. What I discovered that if you guard `home `route (as is required in my app -- no path is public) then it stops working correctly (i.e. start looping) if you remove `checkAuth()` call from app.component. I don't know if that's a bug or expected behaviour and I just set it up incorrectly
I attached a zip with sample. If that's more convenient to you I can create a repo

**Comment by grffn** [comment]
I should also add that if you leave `checkAuth()` call it doesn't help either. When you login then logout then login again, it gets stuck in a loop

**Comment by FabianGosebrink** [comment]
A `package.json` would be fine. Could you add it to the repo? A repo would be great then we can compare the fix if we create one.

**Comment by grffn** [comment]
https://github.com/grffn/demo-auto-login
Please, look

**Comment by FabianGosebrink** [comment]
Have done a PR where the autologin works at my place

**Comment by JerrettDavis** [comment]
Just a small additional data point relating to the `AutLoginGuard`. While I have solved the infinite redirect loop (I think), the guard occasionally produces a bad redirect. I can't seem to reproduce it with any regularity, but it happens occasionally during live reloads.

I get something along the lines of this:
![image](https://user-images.githubusercontent.com/2610199/116746529-3867ac00-a9c2-11eb-9dc8-3f29c0e89c34.png)

Here are the applicable routes:
```typescript
const appRoutes: Routes = [
  {path: '', pathMatch: 'full', redirectTo: 'home'},
  {
    path: 'home',
    canActivate: [AutoLoginGuard],
    loadChildren: () => import('./modules/home/home.module').then(m => m.HomeModule)
  }
  {path: '**', component: PageNotFoundComponent},  // Wildcard route for a 404 page
];
````

All the `AutoLoginGuard`s are present only in the `app-routing.module`, the guard isn't necessary in the child routes since it's secured at the root. 

@FabianGosebrink please chime in if this should be a separate issue.

I can attempt to build a reproducible example, if needed, but maybe the answer is obvious based on just the above information.

**Comment by marcelbarner** [comment]
I also noticed a infinity loop in some special cases. The reproducing steps are:
1. Call a route, which is secured by the `AutoLoginGuard`
2. Now you should be redirected to the Login, do not login.
3. Now call the secured route again
4. Now the loop is started

I will try to offer you a min repro with this behavior.

What I noticed is that we also have very much calls to the authorize-url and values in the storage will be updated continuesly.

**Comment by grffn** [comment]
Sorry for the long delay
I made a commit upon Fabian's modification which makes the issue reproducible for me again
steps to reproduce are the same as in https://github.com/damienbod/angular-auth-oidc-client/issues/1052#issuecomment-839451528

**Comment by grffn** [comment]
@FabianGosebrink sorry to bother you but are there any news? Have you been able to reproduce the issue?

**Comment by FabianGosebrink** [comment]
Hey, thanks for providing a demo. I did not take a look into it so far. Have to give it a shot in the next days. Life is busy currently 😉

**Comment by ebinroy** [comment]
Hi @damienbod @FabianGosebrink ,
    Any solution for this issue. I also stuck here.

**Comment by chad-smith** [comment]
I've spent some time trying to diagnose this, as it was occurring quite readily for us.

The first clue I came across was that it didn't occur in earlier versions, so it could be traced back to an addition after 11.6.2

The second clue came when I finally tweaked the code enough to see where the loop was occuring(<sup>1</sup>). This appeared to be along the lines of:

1. Visit `/secure`
2. User is not authenticated. AutoLoginGuard:
  a. Calls checkAuth (return false)
  b. Stores redirect url ('redirect' in localStorage)
  c. Takes you to the auth service
3. User cancels the auth request or it fails for some reason
4. User visits `/secure` again
5. User is still not authenticated. 
6. AutoLoginGuard calls checkAuth
  a. [Recently added code in checkAuth](https://github.com/damienbod/angular-auth-oidc-client/commit/e4f4a06659262a20ba1bd96ce98543462722f85e#diff-8272aa98e2813e725a83512ddfe1de4fccda27f01c3d105661a39202ad1a1685R73-R79) picks up the redirect url ('/secure') and redirects you
  b. GOTO 4

Incidentally this was occurring readily for us because we were hitting [an issue](https://github.com/damienbod/angular-auth-oidc-client/issues/1100) with the silent renew request getting mixed up with a normal code flow request when we dropped our renew timeout down to a lower value to test it - this caused our authentication request to return a HTTP 400, eventually triggering the redirect loop. 

<sup>1</sup>. Debugging this was a pain because it triggered a `location.href` change in `RedirectService`, which caused the debugger to become ineffective, commenting this out allowed me to get further.

**Comment by FabianGosebrink** [comment]
Hey, Thanks for digging into this. We are working hard behind the scenes, but that information is very useful, thank you. You just had the error case. Is that issue only occurring in error case, so when the user fails to authenticate?

**Comment by chad-smith** [comment]
Hi Fabian,

Yes, as far as I'm aware it only does so on failure. The redirect is cleared when successful, but it's very hard to trace this with the redirects happening and rxjs obfuscating all the stack traces.

We were causing the auth failure by meddling with the silent renew process, as detailed in the linked issue. Depending how your auth service deals with that scenario it might be an easy way for you to force a repro. Simply cancelling the authentication attempt does cause a flood of `authorize` calls for us, but strangely it recovers after the browser throttles requests.

I've just noticed a flaw in my steps above; according to the code, the redirect url is cleared once used, but there might be some interplay with the auto login guard. For me the loop became easier to understand when I commented out the redirect, but that involves copying the source code into our project and I forgot to create a branch of the code in that state.

If the above information doesn't help I can get my environment back into that state and try to provide more details; I switched to looking for workarounds when I started to understand it enough. (The workaround I came up with isn't ideal and is very specific to our app).

**Comment by ebinroy** [comment]
Hi @damienbod @FabianGosebrink ,
 My[ issue# 1099](https://github.com/damienbod/angular-auth-oidc-client/issues/1099) as is same the above but some times it goes infinite loop and sometimes not.

**Comment by FabianGosebrink** [comment]
We will take a look into this as soon as we have time. This is hard to reproduce, but the info are getting us close.

**Comment by robinvanderknaap** [comment]
I was able to reproduce this problem with the `sample-code-flow-auto-login` project. If the call to `checkAuth` in app.component.ts is commented out, you can follow the steps as described by @chad-smith. 

So, basically:
- Visit home page
- Navigate to /protected page
- Do not login!
- Visit home page again 
- Navigate to /protected again, and keep the network tab open, you can see the authorize calls flooding in.

If you remove the redirect key in local storage. You can visit the protected page again, and the problem disappears.

**Comment by andrew-landsverk-win** [comment]
@robinvanderknaap This is exactly my experience, removing the `redirect` key made the problem go away.

**Comment by FabianGosebrink** [comment]
Hey @robinvanderknaap , thanks for mentioning this. If we have the initial route protected by the `AutoLoginGuard`, then the `checkAuth()` is called by the Guard. If the route is open, you need to call `checkAuth()` for yourself.

So when you comment out the `checkAuth()` in your example, you have to add the `AutoLoginGuard` to the first route. I am not sure if you did this or I can not read this in your steps.

Thanks

**Comment by robinvanderknaap** [comment]
Hi Fabian,

Thanks for your response. I'm aware of the need to call checkAuth() on the public route. I just mentioned this, because I think it shows the root cause of this issue. Let me explain:

In the sample project (sample-code-flow-auto-login), when the checkAuth() is not commented, the following behavior occurs: When visiting the home page, you can click the link to the protected page, and the user is being redirected to the authorization server. When the user does not login, but instead navigates to the home page again or to any other public route, the user is immediately being redirected to the authorization server again. So, the home page and all other public routes are not reachable anymore.

The reason for this is the setting of the redirect key in the local storage. The call to checkAuth() always redirects to that page, even if the user is not authenticated. Once the redirect is set in local storage,  every route gets redirected to the authorization server due to the CheckAuth() call in app.component.ts.

For this reason, I and I think others, have setup a specific path in our apps to which the authorization server redirects. So, we don't need the checkAuth call, and simply set the AutoLoginGuard on that specific path.
This is when the problem described in this issue occurs. The AutoLoginGuard also calls the checkAuth method, and that method is still redirecting to the page set in LocalStorage even when the user is not authenticated. This happens before the AutoLoginGuard is able to finish so it seems, and the due to the redirects, we have an infinite loop.

I think the problem lies here in the source code: https://github.com/damienbod/angular-auth-oidc-client/blob/main/projects/angular-auth-oidc-client/src/lib/check-auth.service.ts#L75

The user is always redirected, also when not authenticated.

I may have a solution: Add a authentication check before redirecting:

```
tap(() => {

        const isAuthenticated = this.authStateService.areAuthStorageTokensValid();

        if(isAuthenticated){

          const savedRouteForRedirect = this.autoLoginService.getStoredRedirectRoute();

          if (savedRouteForRedirect) {
            this.autoLoginService.deleteStoredRedirectRoute();
            this.router.navigateByUrl(savedRouteForRedirect);
          }
        }
      }),
```

This fix works in the sample project. CheckAuth() can be left uncommented, and the homepage is always reachable, with the redirect key in local storage. 

It seems logical to me the user needs to be authenticated before being redirected, but maybe I'm missing something, I don't know the library that well.

**Comment by FabianGosebrink** [comment]
Hey thanks for that response! That makes totally sense. Would you like to do a PR for that?

**Comment by chad-smith** [comment]
I'd come to a similar conclusion that making the code in checkAuth more defensive might solve it, but it concerned me that functionality belonging to the AutoLoginGuard was creeping further into other areas.

**Comment by robinvanderknaap** [comment]
> Hey thanks for that response! That makes totally sense. Would you like to do a PR for that?

No problem

**Comment by LiuCypher** [comment]
Hey there! Thanks a lot! The fix worked fine for us and we are looking into using it in production, do you know when it will be available? thanks again :)

**Comment by damienbod** [comment]
I plan to test and release tomorrow 11.6.10

**Comment by FabianGosebrink** [comment]
Thanks @robinvanderknaap for the PR!

**Comment by damienbod** [comment]
released in 11.6.10
